### PR TITLE
2.2.x, pac4j 1.8

### PR DESCRIPTION
**Pull request by andriykuba**

### CODE REVIEW

**Comment by leleuj** [comment]
Great! It looks very good to me. @nielsdraaisma : any feedback on the PR?

**Comment by leleuj** [comment]
I created the 2.2.x branches on both demos to update them with the play-pac4j 2.2.x branch (after the merge).

File: pom.xml [hunk 1]
```diff
@@ -73,7 +73,7 @@
 
     <properties>
         <pac4j.version>1.8.6</pac4j.version>
```
Comment (leleuj) [inline pom.xml]: By the way, we can upgrade to pac4j 1.8.8.

**Comment by nielsdraaisma** [comment]
This looks pretty similar to the work I've done in my fork a while ago, would be great to get this in.

**Comment by leleuj** [comment]
I just merged in and I also updated the README (master and 2.2.x) to make things clearer.

@andriykuba : would you mind updating the 2.2.x branches of the demos (https://github.com/pac4j/play-pac4j-java-demo/tree/2.2.x & https://github.com/pac4j/play-pac4j-scala-demo/tree/2.2.x) with the new 2.2.0-SNAPSHOT version of play-pac4j? Thanks

**Comment by andriykuba** [comment]
@leleuj: I will look. 

In a short - in the case of `HttpConstants.TEMP_REDIRECT` the `org.pac4j.play.PlayWebContext.getResponseLocation()` returns null.

But `org.pac4j.play.PlayWebContext.getJavaContext().response().getHeaders()` has the correct location header. 

I found that I changed the test `testRedirectToOriginallyRequestedUrl()` from 
`assertEquals("Location must be equal to " + expected, expected, result.header("Location"));`
to 
`assertEquals("Location must be equal to " + expected, expected, result.header("Location").get());`

As I see that was wrong. So please, hold on with the release, I will figure it out.

**Comment by andriykuba** [comment]
@leleuj: I found the reason of this problem:
* `PlayWebContext` inherits `BaseResponseContext` 
* `BaseResponseContext`  has  `getResponseLocation` method 

```java
...
protected final Map<String, String> responseHeaders = new HashMap<>();        
...

public String getResponseLocation() {
  return this.responseHeaders.get(HttpConstants.LOCATION_HEADER);
}
```

* `PlayWebContext` does not have this method and I cannot figure out how it could set the `responseHeaders` of the parrent class because it shadows the `setResponseHeader` method

```java
...
protected final Response response;
...

@Override
public void setResponseHeader(final String name, final String value) {
  response.setHeader(name, value);
}
```

* So I am wandering how it works in a previous version. 
* I propose the "fall back" fix to the `setResponseHeader` method of the `PlayWebContext` class:

```java
@Override
public String getResponseLocation() {
  String location = super.getResponseLocation();
  if(location == null){
    location = response.getHeaders().get(HttpConstants.LOCATION_HEADER);
  }
  return location;
}
```

With this fix all test was passed and "play-pac4j-java-demo" works correctly

**Comment by leleuj** [comment]
Apart from the test, do the demos work with the current code?

Because it's normal to have the location via the `getResponseLocation` method as it's used by the `DefaultHttpActionAdapter`:  https://github.com/pac4j/play-pac4j/blob/master/src/main/java/org/pac4j/play/http/DefaultHttpActionAdapter.java#L32

So I would keep the code as is, check the demos work and update (or remove) the test accordingly.

**Comment by andriykuba** [comment]
Hello @leleuj, I spent some time for the deeper look. 

> Apart from the test, do the demos work with the current code?

Tests are not the issue, they are correct. That was a glance, and now I see that they have nothing to this issue. 

No, demos does not work in the current code. (and yes, test are correct, they just not cover this problem)

>Because it's normal to have the location via the getResponseLocation method 

Yes, but this method works incorrectly for the Play. I tried to explain this in my previous comment. Here is the more detailed explanation.

I found that the **2.1.0** version of **play-pac4j** have the same issue: 

> In the case of 'HttpConstants.TEMP_REDIRECT' the 'org.pac4j.play.PlayWebContext.getResponseLocation()' returns null

 It obvious from the code but I did verification on the live program. 

The reason that example works in the **2.1.0** with **Play 2.4.x** is that in **Play 2.4.x** `redirect((String) null);` does not throw an exception, but in **Play 2.5.x** it throws the `java.lang.NullPointerException: Response header 'Location' has null value!` exception, so example fails. 

And another question why then it works in **2.1.0** if it redirects to `null`? 

In the `org.pac4j.play.java.RequiresAuthenticationAction.redirectToIdentityProvider`, actually before action adapter is called there is a code

    final IndirectClient currentClient = (IndirectClient) currentClients.get(0);
    currentClient.redirect(context, true);

And in the `IndirectClient.redirect` the code is 
	
	...
	context.setResponseHeader(HttpConstants.LOCATION_HEADER, action.getLocation());
	...

So, before it calls to `redirect((String)null)`, the header is already set. Yeh, look on my previous post, `PlayWebContext` sets the header into the Play response object correctly.

Play put more priority on the response headers than parameter of the redirect function. For example the next code will redirect to the "http://www.google.com":

      ctx().response().setHeader("Location", "http://www.google.com");
      return redirect("https://www.bing.com");

So it's completely clean that `PlayWebContext` must be fixed for the **Play 2.5.x**. 

If you strictly do not want to change the code of **pac4j-play 2.2.x** then, as a temporary solution for the examples, I can just implement `DemoHttpActionAdapter` as a `HttpActionAdapter` (actually mix the current `DemoHttpActionAdapter` witht the default `DefaultHttpActionAdapter`) and use some string as the `redirect` parameter to avoid the NullPointerException. Like this:

    ...
    } else if (code == HttpConstants.TEMP_REDIRECT) {
      String location = ((PlayWebContext)context).getJavaContext().response().getHeaders().get(HttpConstants.LOCATION_HEADER))
      return redirect(location);
    ...

**Comment by leleuj** [comment]
It works on both demos for branch 2.1.x, but following your explanations, it seems be a luck more than anything else.
And with the change in Play 2.5 (exception when redirection is null), it doesn't. And I admit that things are a bit messy between all components: `PlayWebContext`, `IndirectClient.redirect`, `HttpActionAdapter`...

I'm in line with you: my feeling is that the easiest solution would be to change the `DefaultHttpActionAdapter`. Can you go that way?

**Comment by andriykuba** [comment]
There are requests 

https://github.com/pac4j/play-pac4j/pull/110
https://github.com/pac4j/play-pac4j-java-demo/pull/32
https://github.com/pac4j/play-pac4j-scala-demo/pull/30

**Pull request merged**